pkg/sql/opt_index_selection.go

```diff
@@ -613,18 +630,70 @@ func spanFromConstraintSpan(
 	// Encode each logical part of the end key.
 	s.EndKey, err = encodeConstraintKey(index, cs.EndKey(), interstices)
 	if err != nil {
-		return roachpb.Span{}, err
+		return nil, err
 	}
 	s.EndKey = append(s.EndKey, interstices[cs.EndKey().Length()]...)
 
+	// Optimization: for single row lookups on a table with multiple column
+	// families, only scan the relevant column families. This is disabled for
+	// deletions to ensure that the entire row is deleted.
+	if !forDelete &&
+		needed.Len() > 0 &&
+		index.ID == tableDesc.PrimaryIndex.ID &&
+		len(tableDesc.Families) > 1 &&
+		cs.StartKey().Length() == len(tableDesc.PrimaryIndex.ColumnIDs) &&
+		s.Key.Equal(s.EndKey) {
```

Comment:
Is there a way we can avoid this bytes comparison? I can't imagine it's that expensive but I'm surprised there's no way to detect this condition without having to actually compare bytes, since we have all of the logical information a little above.

Thinking about this in context of #22219, which has a similar need - if we can remember the expected arity of a scan, we can improve performance of various internals further down in the stack.

Reply:
I don't see a way to determine this from the `constrant.Span` without doing a similar comparison. It seems like we would need to add a flag indicating that it's a single row lookup and teach `constraintsBuilder` to set this flag appropriately, which hopefully would be pretty straightforward, but outside the scope of this PR.

Reply:
Cool, sounds good to me.